assets/src/block-editor/setupBlockEditorValidation.js

```diff
@@ -0,0 +1,182 @@
+    }
+
+    const invalidBlocks = blocks.reduce((invalidBlocksArray, block) => {
+      // Normally `blocks` contains a valid list of blocks, however it can happen that one of them is `null` in rare
+      // cases. It happened to me once while running with WordPress 5.8.1 and undoing multiple edits. This made the
+      // editor crash while it's trying to access `block.name`.
+      if (!block) {
+        return; // eslint-disable-line array-callback-return
+      }
+      const validations = blockValidations[block.name] || {};
+
+      const results = Object.entries(validations).reduce((resultsArray, [attrName, validate]) => {
+        const value = block.attributes[attrName];
+        const result = validate(value);
+
+        if (!result.isValid) {
+          resultsArray.push(result);
+        }
+
+        return resultsArray;
+      }, []);
+
+      invalidBlocksArray.push(...results);
+
+      return invalidBlocksArray;
+    }, []);
+
+    // Push invalid block validation messages to currentMessages
+    invalidBlocks.forEach(block => currentMessages.push(...block.messages));
+
+    // Determine if the post is currently valid (no validation errors)
+    const currentlyValid =
+      (0 === invalidBlocks.length) &&
+      (0 === topicLinkBlocksNeedsImage.length) &&
+      !invalidTitle &&
+      !needsFeaturedImage;
+
+    // Update the global messages array with the current validation messages
+    messages = currentMessages;
+
+    // Check if the validation status has changed
+    if (canPublish === currentlyValid) {
+      return;
+    }
+    canPublish = currentlyValid;
+
+    // Lock or unlock post saving based on validity
+    if (!canPublish) {
+      dispatch('core/editor').lockPostSaving();
+    } else {
+      dispatch('core/editor').unlockPostSaving();
+    }
+  });
+
+  // Register the pre-publish checklist panel plugin
+  registerPlugin('pre-publish-checklist', {render: PrePublishCheckList});
+
+  // Add filter for block type validation
+  wp.hooks.addFilter(
+    'blocks.registerBlockType',
+    'planet4-plugin-gutenberg-blocks',
+    registerAttributeValidations
+  );
+};
+
+/**
+ * Event listener for changes in the campaign name select field.
+ * Updates the post meta with the selected campaign name.
+ */
+document.addEventListener('change', e => {
+  if (!e.target.matches('select[name="p4_campaign_name"]')) {
+    return;
+  }
+  dispatch('core/editor').editPost({meta: {p4_campaign_name: e.target.value}});
+});
+
+/**
+ * Registers block attribute validations.
+ * @param {Object} settings - The block settings object.
+ * @param {string} blockName - The name of the block.
+ * @return {Object} - The updated block settings with validations.
+ */
+const registerAttributeValidations = (settings, blockName) => {
+  const {attributes} = settings;
+
+  // Register block attribute validations
+  Object.keys(settings.attributes).forEach(attrName => {
```

Comment:
What if you use `attributes` instead `settings.attributes` if you've already destructured?

Reply:
This file was moved from the P4 plugin and only lines 49 to 57 and 87 to 92 were modified.
If you don't mind, as your question is out of the scope of this task, I'd rather fix it on another PR.